sdks/python/apache_beam/transforms/periodicsequence.py

```diff
@@ -0,0 +1,165 @@
+    return restriction.size()
+
+
+class ImpulseSeqGenRestrictionTracker(OffsetRestrictionTracker):
+  def try_split(self, fraction_of_remainder):
+    if not self._checkpointed:
+      if fraction_of_remainder != 0:
+        return None
+
+      if self._current_position is None:
+        cur = self._range.start
+      else:
+        cur = self._current_position
+      split_point = cur
+
+      if split_point < self._range.stop:
+        self._checkpointed = True
+        self._range, residual_range = self._range.split_at(split_point)
+        return self._range, residual_range
+
+  def cur_pos(self):
+    return self._current_position
+
+  def try_claim(self, pos):
+    if ((self._last_claim_attempt is None) or
+        (pos > self._last_claim_attempt and pos == self._range.stop)):
+      self._last_claim_attempt = pos
+      return True
+    else:
+      return super(ImpulseSeqGenRestrictionTracker, self).try_claim(pos)
+
+
+class ImpulseSeqGenDoFn(beam.DoFn):
+  def process(
+      self,
+      element,
+      restriction_tracker=beam.DoFn.RestrictionParam(
+          ImpulseSeqGenRestrictionProvider())):
+    '''
+    PeriodicSequence transform receives tuple elements with three parts:
+
+    * first_timestamp = first timestamp to output element for.
+    * last_timestamp = last timestamp/time to output element for.
+    * fire_interval = how often to fire an element.
+
+    For each input element received, PeriodicSequence transform will start
+    generating output elements in following pattern:
+
+    * if element timestamp is less than current runtime then output element.
+    * if element timestamp is greater than current runtime, wait until next
+      element timestamp.
+
+    PeriodicSequence can't guarantee that each element is output at exact time.
+    PeriodicSequence guarantees that elements would not be output prior to given
+    runtime timestamp.
+
+    :param element: (start_timestamp, end_timestamp, interval)
+    :param restriction_tracker:
+    :return: yields elements at processing real-time intervals with value of
+      target output timestamp for the element.
+    '''
+    _, _, interval = element
+
+    assert isinstance(restriction_tracker, sdf_utils.RestrictionTrackerView)
+
+    t = time.time()
+    cr = restriction_tracker.current_restriction()
+    current_timestamp = cr.start
+
+    restriction_tracker.try_claim(current_timestamp)
+    if current_timestamp <= t:
+      if restriction_tracker.try_claim(current_timestamp + interval):
+        current_timestamp += interval
+        yield current_timestamp
+
+    if current_timestamp + interval >= cr.stop:
+      restriction_tracker.try_claim(cr.stop)
+    else:
+      restriction_tracker.defer_remainder(
+          timestamp.Timestamp(current_timestamp))
+
+
+class PeriodicSequence(PTransform):
+  """
+  See ImpulseSeqGenDoFn.
+  """
+  def __init_(self):
+    pass
+
+  def expand(self, pbegin):
+    return (
+        pbegin
```

Comment:
Renaming to pcoll, as the input is an actual Pcollection, right?

```suggestion
  def expand(self, pcoll):
    return (
        pcoll
```